File: javascript/src/composite/composite.js
/** Stores children and delegates the operation to them; child management lives here (safe variant). */
export class Composite {
  #children = [];

  add(child) {
    this.#children.push(child);
    return this;
  }

  operation() {
    return `Composite(${this.#children.map((c) => c.operation()).join("+")})`;
  }
}

File: java/src/main/java/com/penapereira/patterns/composite/Composite.java
package com.penapereira.patterns.composite;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/** Stores children and delegates the operation to them; child management lives here (safe variant). */
public final class Composite implements Component {

    private final List<Component> children = new ArrayList<>();

    public Composite add(Component child) {
        children.add(child);
        return this;
    }

    @Override
    public String operation() {
        return "Composite(" + children.stream().map(Component::operation).collect(Collectors.joining("+")) + ")";
    }
}
